core/src/test/java/com/linecorp/armeria/client/EndpointTest.java

```diff
@@ -445,4 +452,82 @@ void inetAddress() throws UnknownHostException {
         assertThat(endpointWithIpv6.ipFamily()).isEqualTo(StandardProtocolFamily.INET6);
         assertThat(endpointWithIpv6.ipAddr()).isEqualTo("0:0:0:0:0:0:0:1");
     }
+
+    @Test
+    void setAndGetAttr() {
+        final Endpoint endpointA = Endpoint.parse("a");
+
+        final AttributeKey<String> key1 = AttributeKey.valueOf("key1");
+        final AttributeKey<String> key2 = AttributeKey.valueOf("value2");
+        final AttributeKey<Object> objKey = AttributeKey.valueOf("objKey");
+        final Object objValue = new Object();
+        final Endpoint endpointB = endpointA.withAttr(key1, "value1")
+                                            .withAttr(key2, "value2")
+                                            .withAttr(objKey, objValue)
+                                            .withAttr(key1, "value1-1");
+
+        assertThat(endpointB).isNotSameAs(endpointA);
+        assertThat(endpointA.attr(key1)).isNull();
+        assertThat(endpointB.attr(key1)).isEqualTo("value1-1");
+        assertThat(endpointB.attr(key2)).isEqualTo("value2");
+
+        // key with same value
+        assertThat(endpointB.withAttr(objKey, objValue)).isSameAs(endpointB);
+        assertThat(endpointB.withAttr(AttributeKey.valueOf("keyNotFound"), null)).isSameAs(endpointB);
+
+        // value remove
+        final Endpoint endpointC = endpointB.withAttr(AttributeKey.valueOf("key1"), null);
+        assertThat(endpointC).isNotSameAs(endpointB);
+        assertThat(endpointC.attr(key1)).isNull();
+    }
+
+    @Test
+    void attrs() {
+        final Endpoint endpoint = Endpoint.parse("a");
+        assertThat(endpoint.attrs()).isExhausted();
+
+        final List<Entry<AttributeKey<?>, String>> attrs = new ArrayList<>();
+        final AttributeKey<String> key1 = AttributeKey.valueOf("key1");
+        final AttributeKey<String> key2 = AttributeKey.valueOf("key2");
+
+        attrs.add(new AbstractMap.SimpleImmutableEntry<>(key1, "value1"));
```

Comment:
nit: Could use Guava's API for the sake of simplicity.
```suggestion
        attrs.add(Maps.immutableEntry(key1, "value1"));
```

Reply:
It's updated. 😃